add blog for how to deploy LLM with inference endpoints (#1286)
huggingface/blog: inference-endpoints-llm.md

---
title:  Deploy LLMs with Hugging Face Inference Endpoints
thumbnail: /blog/assets/155_inference_endpoints_llm/thumbnail.jpg
authors:
- user: philschmid
---

# Deploy LLMs with Hugging Face Inference Endpoints

<!-- {blog_metadata} -->
<!-- {authors} -->

Open-source LLMs like [Falcon](https://huggingface.co/tiiuae/falcon-40b), [(Open-)LLaMA](https://huggingface.co/openlm-research/open_llama_13b), [X-Gen](https://huggingface.co/Salesforce/xgen-7b-8k-base), [StarCoder](https://huggingface.co/bigcode/starcoder) or [RedPajama](https://huggingface.co/togethercomputer/RedPajama-INCITE-7B-Base), have come a long way in recent months and can compete with closed-source models like ChatGPT or GPT4 for certain use cases. However, deploying these models in an efficient and optimized way still presents a challenge.

> In this blog post, we will show you how to deploy open-source LLMs to [Hugging Face inference Endpoints](https://ui.endpoints.huggingface.co/), our managed SaaS solution that makes it easy to deploy models. Additionally, we will teach you how to stream responses and test the performance of our endpoints. So let's get started!

Reviewer: ```suggestion
In this blog post, we will show you how to deploy open-source LLMs to [Hugging Face Inference Endpoints](https://ui.endpoints.huggingface.co/), our managed SaaS solution that makes it easy to deploy models. Additionally, we will teach you how to stream responses and test the performance of our endpoints. So let's get started!
```

2. [Cost Efficiency](https://huggingface.co/docs/inference-endpoints/autoscaling): Benefit from automatic scale to zero capability, reducing costs by scaling down the infrastructure when the endpoint is not in use, while paying based on the uptime of the endpoint, ensuring cost-effectiveness.
3. [Enterprise Security](https://huggingface.co/docs/inference-endpoints/security): Deploy models in secure offline endpoints accessible only through direct VPC connections, backed by SOC2 Type 2 certification, and offering BAA and GDPR data processing agreements for enhanced data security and compliance.
4. [LLM Optimization](https://huggingface.co/text-generation-inference): Optimized for LLMs, enabling high throughput with Paged Attention and low latency through custom transformers code and Flash Attention power by Text Generation Inference
5. [Wide Task Support](https://huggingface.co/docs/inference-endpoints/supported_tasks): Support for 🤗 Transformers, Sentence-Transformers, Diffusion tasks, and custom tasks like speaker diarization, expanding deployment possibilities.

> You can get started with inference endpoints at: [https://ui.endpoints.huggingface.co/](https://ui.endpoints.huggingface.co/)

Reviewer: ```suggestion
You can get started with Inference Endpoints at: [https://ui.endpoints.huggingface.co/](https://ui.endpoints.huggingface.co/)
```

## 1. How to deploy Falcon 40B instruct

> Before you can get started, make sure to meet all of the following requirements:
> 
> 1. An Organization/User with an active credit card. (Add billing **[here](https://huggingface.co/settings/billing)**)
> 2. You can access the UI at: **[https://ui.endpoints.huggingface.co](https://ui.endpoints.huggingface.co/endpoints)**

Reviewer: ```suggestion
To get started, you need to be logged in with a User or Organization account with a payment method on file (you can add one **[here](https://huggingface.co/settings/billing)**), then access Inference Endpoints at **[https://ui.endpoints.huggingface.co](https://ui.endpoints.huggingface.co/endpoints)**
```

> Once you are logged in, you can click on “New endpoint”. Select the repository, the cloud, and the region, adjust the instance and security settings, and deploy in our case `tiiuae/falcon-40b-instruct`.

Reviewer: ```suggestion
Then, click on “New endpoint”. Select the repository, the cloud, and the region, adjust the instance and security settings, and deploy in our case `tiiuae/falcon-40b-instruct`.
```

![Select Hugging Face Repository](assets/assets/155_inference_endpoints_llm/repository.png "Select Hugging Face Repository")

Inference Endpoints suggest an instance type for based on the model size, which should be big enough to run the model. Here `4x NVIDIA T4` GPUs. To get the best performance for the LLM, change the instance to `GPU [xlarge] · 1x Nvidia A100`.

*Note: If the instance type cannot be selected, you need to [contact us](mailto:api-enterprise@huggingface.co?subject=Quota%20increase%20HF%20Endpoints&body=Hello,%0D%0A%0D%0AI%20would%20like%20to%20request%20access/quota%20increase%20for%20{INSTANCE%20TYPE}%20for%20the%20following%20account%20{HF%20ACCOUNT}.) and request an instance quota.*

![Select Instance Type](assets/assets/155_inference_endpoints_llm/instance-selection.png "Select Instance Type")

> You can then deploy our model by clicking “Create Endpoint”. After 10 minutes, the Endpoint should be online and available to serve requests. 

Reviewer: ```suggestion
You can then deploy your model with a click on “Create Endpoint”. After 10 minutes, the Endpoint should be online and available to serve requests. 
```

## 2. Test the LLM endpoint

> The Endpoint overview provides access to the Inference Widget, which can be used to send requests. This allows you to quickly test your Endpoint with different inputs and share it with team members. Those Widgets don’t support parameters, which results, in this case, a “short” generation. 

Reviewer: ```suggestion
The Endpoint overview provides access to the Inference Widget, which can be used to manually send requests. This allows you to quickly test your Endpoint with different inputs and share it with team members. Those Widgets do not support parameters - in this case this results to a “short” generation. 
```

![Test Inference Widget](assets/assets/155_inference_endpoints_llm/widget.png "Test Inference Widget")

The widget also generates a cURL command you can use. Just add your `hf_xxx` and test. 

```python
curl https://j4xhm53fxl9ussm8.us-east-1.aws.endpoints.huggingface.cloud \
-X POST \
-d '{"inputs":"Once upon a time,"}' \
-H "Authorization: Bearer <hf_token>" \
-H "Content-Type: application/json"
```

> You can use different parameters to control the generation, defining them in the `parameters` attribute of the payload. As of the following parameters are supported:

Reviewer: ```suggestion
You can use different parameters to control the generation, defining them in the `parameters` attribute of the payload. As of today, the following parameters are supported:
```

- `stop`: A list of tokens to stop the generation. The generation will stop when one of the tokens is generated.
- `top_k`: The number of highest probability vocabulary tokens to keep for top-k-filtering. Default value is `null`, which disables top-k-filtering.
- `top_p`: The cumulative probability of parameter highest probability vocabulary tokens to keep for nucleus sampling, default to `null`
- `do_sample`: Whether or not to use sampling; use greedy decoding otherwise. Default value is `false`.
- `best_of`: Generate best_of sequences and return the one if the highest token logprobs, default to `null`.
- `details`: Whether or not to return details about the generation. Default value is `false`.
- `return_full_text`: Whether or not to return the full text or only the generated part. Default value is `false`.
- `truncate`: Whether or not to truncate the input to the maximum length of the model. Default value is `true`.
- `typical_p`: The typical probability of a token. Default value is `null`.
- `watermark`: The watermark to use for the generation. Default value is `false`.

## 3. Stream responses in Javascript and Python

> Requesting and generating text with LLMs can be a time-consuming and iterative process. To improve the user experience by streaming tokens as they are generated. Below are two examples of how to stream tokens using Python and JavaScript. For Python, we are going to use the [client from Text Generation Inference](https://github.com/huggingface/text-generation-inference/tree/main/clients/python), and for JavaScript, the [HuggingFace.js library](https://huggingface.co/docs/huggingface.js/main/en/index)

Reviewer: ```suggestion
Requesting and generating text with LLMs can be a time-consuming and iterative process. A great way to improve the user experience is streaming tokens to the user as they are generated. Below are two examples of how to stream tokens using Python and JavaScript. For Python, we are going to use the [client from Text Generation Inference](https://github.com/huggingface/text-generation-inference/tree/main/clients/python), and for JavaScript, the [HuggingFace.js library](https://huggingface.co/docs/huggingface.js/main/en/index)
```

### Streaming requests with Python

First, you need to install the `text-generation` client

```python

> pip install text-generation

Reviewer 2: Important point, i would have rather showcased `huggingface_hub` instead of `text-generation` as it contains  the exact same code

Better to focus usage / visibility on one client library for critical mass

cc @OlivierDehaene @Wauplin @philschmid

Reviewer 3: Yes this is new @philschmid and is shipped today. I'll make a PR and tag you :)

Reviewer 3: Opened a PR: https://github.com/huggingface/blog/pull/1290

Reviewer 3: Merged.

pip install text-generation
```

We can create a `Client` providing our endpoint URL and credential alongside the hyperparameter we want to use

```python
from text_generation import Client

> # HF Inference endpoints parameter

Reviewer: ```suggestion
# HF Inference Endpoints parameter
```
